### PR TITLE
fix: Skip cached public keys and cached shared keys from enrollment a…

### LINKED ISSUE #1335: Add unit tests to verify enrollment authorization check during put and get operation in at_client
### Is your feature request related to a problem? Please describe.

When an atSign undergoes authentication via the APKAM, the system verifies the key's namespace to ensure it's authorized for enrollment during both "put" and "get" operations.  In the process of enrollment authorization check the following key-types are excluded : Reserved Keys, cached shared key and cached public key. 

Currently there no unit tests in place to validate this behavior.

### Describe the solution you'd like

Add unit tests to assert the enrollment authorization changes and the mentioned key types should be excluded from the enrollment authorization check.

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
…uthorization check

**- What I did**
 - When authenticated with APKAM, operations performed in a local secondary database exclude only reserved keys from validation. However, there is an issue with validating the cached `sharedkey`and `encryption public key`because these keys lack a namespace for validation.
 - Along with the reserved keys, exclude the cached shared_key and cached public key from the enrollment authorization check.

**- How I did it**
In local_secondary.dart, modify the existing method which excludes the local keys from enrollment authorization check to also exclude the cached `sharedkey` and `encryption public key`.

**- How to verify it**
- All the existing unit and functional test should pass.
- Tested the no-ports E2E tests locally with APKAM authentication.

Attaching the test logs
[srvd.log](https://github.com/user-attachments/files/15787816/srvd.log)
[sshnp.log](https://github.com/user-attachments/files/15787817/sshnp.log)
[sshnpd.log](https://github.com/user-attachments/files/15787818/sshnpd.log)

Pending work:
- Add unit tests to ensure the cached shared key, cached public key and reserved keys are excluded from enrollment authorization check. Created a git issue and working on it : https://github.com/atsign-foundation/at_client_sdk/issues/1335

### CODE REVIEW

**Comment by murali-shris** [comment]
is there any test which covers these changes?

**Comment by sitaram-kalluri** [comment]
> is there any test which covers these changes?

Created a git issue, working on the changes.

**Comment by sitaram-kalluri** [comment]
> > is there any test which covers these changes?
> 
> Created a git issue, working on the changes.

Completed adding of unit tests.

File: packages/at_client/test/test_utils/test_utils.dart [hunk 1]
```diff
@@ -1,10 +1,40 @@
 import 'dart:math';
 
+import 'package:at_chops/at_chops.dart';
+
 class TestUtils {
```
Comment (gkc) [inline packages/at_client/test/test_utils/test_utils.dart:5]: Instead of hard-coding keys, please generate them during test setup

**Review by gkc — CHANGES_REQUESTED** [review]
Please add tests which explicitly assert behaviour of LocalSecondary.isEnrollmentAuthorizedForOperation for a variety of verbs and keys

e.g. llookup for `cached:@alice:shared_key@bob`, `cached:@alice:key.in.any.name.space@bob`, `cached:@alice:foo@bob` as well as their non-cached equivalents; and also add tests for updates and deletes and notify

I suggest you modify LocalSecondary renaming _enrollment to enrollment and making it `@visibleForTesting` to make it possible to write many small test cases like this
```
    test('test llookup on cached:@alice:shared_key@bob', () async {
      final testEnrollmentId = 'aaa111';

      LocalSecondary ls = 
      <do setup of a LocalSecondary in a test setup function>

      ls.enrollment = Enrollment()
        ..enrollmentId = testEnrollmentId
        ..appName = 'testApkamAuthCachedLLookup'
        ..deviceName = 'testDevice'
        ..namespace = {"fubar": "rw"};

      final bool authorized = await ls.isEnrollmentAuthorizedForOperation(
          'cached:@alice:shared_key@bob', LLookupVerbBuilder());
      expect (authorized, true);
    });

```

File: packages/at_client/test/test_utils/test_utils.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_client/test/test_utils/test_utils.dart:5]: Removed the hard-coded keys. Keys are generated during the running of the test.

**Comment by sitaram-kalluri** [comment]
> Please add tests which explicitly assert behaviour of LocalSecondary.isEnrollmentAuthorizedForOperation for a variety of verbs and keys
> 
> e.g. llookup for `cached:@alice:shared_key@bob`, `cached:@alice:key.in.any.name.space@bob`, `cached:@alice:foo@bob` as well as their non-cached equivalents; and also add tests for updates and deletes and notify
> 
> I suggest you modify LocalSecondary renaming _enrollment to enrollment and making it `@visibleForTesting` to make it possible to write many small test cases like this
> 
> ```
>     test('test llookup on cached:@alice:shared_key@bob', () async {
>       final testEnrollmentId = 'aaa111';
> 
>       LocalSecondary ls = 
>       <do setup of a LocalSecondary in a test setup function>
> 
>       ls.enrollment = Enrollment()
>         ..enrollmentId = testEnrollmentId
>         ..appName = 'testApkamAuthCachedLLookup'
>         ..deviceName = 'testDevice'
>         ..namespace = {"fubar": "rw"};
> 
>       final bool authorized = await ls.isEnrollmentAuthorizedForOperation(
>           'cached:@alice:shared_key@bob', LLookupVerbBuilder());
>       expect (authorized, true);
>     });
> ```

Added unit tests to assert the "isEnrollmentAuthorizedForOperation" method.

File: packages/at_client/lib/src/client/local_secondary.dart [hunk 2]
```diff
@@ -265,42 +266,42 @@ class LocalSecondary implements Secondary {
     return isStored != null ? true : false;
   }
 
+  @visibleForTesting
```
Comment (gkc) [inline packages/at_client/lib/src/client/local_secondary.dart]: This method can continue to be fully visible

Comment (sitaram-kalluri) [reply packages/at_client/lib/src/client/local_secondary.dart]: Removed the annotation and pushed the changes.

**Review by gkc — APPROVED** [review]

**Pull request merged**